### ISSUE TITLE
Support training resume and saving best model for SWA

### ISSUE THREAD

**Issue by b02202050** [Lightning-AI/lightning#6074]
## 🚀 Feature
To support stop-resume and saving the model based on the best validation performance when using Stochastic weight averaging.

### Motivation
If one does not reach the end of the training, the SWA model would not be saved to the checkpoint.
For example:
1. One may want to save the best validation performance model before overfitting.
2. The training becomes unstable to reach the final epoch.

I observe that the saved checkpoint is still the original model even if I use an SWA callback.

### Pitch
1. Use SWA model to run validation step instead of the original model.
    1. We may consider to **Load SWA weight before validation** and **Restore original model weight after validation**
1. Save both original model weight and SWA model weight into checkpoint for resume training.
    1. With the concern of say one want to load the model for testing, he/she might expect the model be the SWA weight. Therefore, we can save the SWA weight to the `checkpotint['state_dict']` and let the original model weight saved as a callback state.

### Additional context
@Borda @MilesCranmer, I do not have much experience on coding neither am I familiar with PyTorch Lightning. would you help? Thx a lot. :wink:

**Comment by adamreeve** [comment]
I've been looking into how batch normalization currently works when using SWA to better understand how this should be handled during validation. The mean and variance of inputs to the BatchNorm layers depend on the weights of lower layers so the running estimates computed by the underlying model won't be accurate when the weights are replaced with average weights. Therefore the mean and variance are computed with a full pass over the training set at the end of training, after replacing the weights with the averaged weights. This is done by increasing the number of epochs by one but setting `_skip_backward` to True for the last epoch to skip the optimization step, and modifying the batch norm layers to reset their mean and variance tensors before this final epoch and setting momentum to None so that they compute a simple cumulative moving average.

The validation passes should probably also update the batch norm parameters in order to accurately represent the model performance. Using the same approach of adding extra non-optimization training epochs before validation seems like it could get quite awkward though. I'm wondering whether it would be simpler to do this without modifying the training loop parameters by just using the training data fetcher directly. I've updated my draft PR to use this approach but I'm not that familiar with the PyTorch code base so am not sure what I might be missing by using the data fetcher directly rather than letting the training loop handle this.

The `StochasticWeightAveraging.restore_average_parameters_from_checkpoint` method will also need to be updated to take a data module or training data loader so that it can recompute batch norm parameters.

**Comment by carmocca** [comment]
Implemented in https://github.com/Lightning-AI/lightning/pull/9938
